sql/import: add row count validation

ZhouXing19 opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/146327

Add optional row count validation for forward indexes after import completion. By default, this validation is enabled, but this means additional row count checks can increase IMPORT statement latency.

The validation verifies that the row count difference for each index matches the expected row count from the IMPORT operation. Partial indexes are also validated.

Release note (sql change): Added cluster setting `bulkio.import.row_count_validation.enabled` (default `true`) to enable row count validation for import operations.

ZhouXing19 committed:
sql: add index row count validation for import

Add optional row count validation for forward indexes after import
completion. By default, this validation is enabled, but this means additional
row count checks can increase IMPORT statement latency.

The validation verifies that the row count difference for each index
matches the expected row count from the IMPORT operation. Partial
indexes are also validated.

Release note (sql change): Added cluster setting
`bulkio.import.row_count_validation.enabled` (default `true`) to enable
row count validation for import operations.

rytaft commented:
Thanks for working on this! I think we'll want to enable this setting by default -- a bit of extra latency should be an acceptable price to pay for correctness validation.





<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice work! I have some questions to consider and some nits.

> @yuzefovich reviewed 2 of 7 files at r1, 2 of 3 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-:-OYDqcKE9wUUepofHXyv:b-f5e8ap)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`pkg/sql/backfill.go` line 2044 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDkhWT5_4-Z8jHlEws:-OYDkhWT5_4-Z8jHlEwt:b-cz1uu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/backfill.go#L2044)):*
> ```Go
> }
> 
> // CountIndexRowsAndMaybeCheckUniqueness returns the number of rows of
> ```

nit: `s/of rows of/of rows in/`.
___
*[`pkg/sql/importer/import_job.go` line 120 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDlAXkCVELXhZUHsDf:-OYDlAXkCVELXhZUHsDg:bdgr307) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L120)):*
> ```Go
> 		"NOTE: this setting could result in latency increase for import jobs",
> 	true,
> 	settings.WithPublic,
> ```

nit: should we also mark it as `settings.Unsafe` (similar to `bulkio.import.constraint_validation.unsafe.enabled`)? I agree with Becca that we want to have this enabled by default, and marking the setting as unsafe will make it even scarier to disable this validation (which seems prudent to me).
___
*[`pkg/sql/importer/import_job.go` line 126 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDm72I9fXRwJCJum3B:-OYDm72J0pfqORZdwrCl:baudemt) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L126)):*
> ```Go
> // before and after import to validate the expected number of rows were imported.
> type expectedRowCountWithSignal struct {
> 	ready            chan struct{}
> ```

nit: why not send the expected row count on the channel? Then we can avoid creating this utility struct.
___
*[`pkg/sql/importer/import_job.go` line 134 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDnmST34Bed4pXu304:-OYDnmST34Bed4pXu305:bw7rksb) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L134)):*
> ```Go
> 	p := execCtx.(sql.JobExecContext)
> 	newHistoricalInternalExecByTime := func(now hlc.Timestamp) descs.HistoricalInternalExecTxnRunner {
> 		return descs.NewHistoricalInternalExecTxnRunner(p.ExecCfg().Clock.Now(), func(ctx context.Context, fn descs.InternalExecFn) error {
> ```

We should probably use `now` instead of `p.ExecCfg().Clock.Now()` so that `now` is used in two places (similar to other callsites of `descs.NewHistoricalInternalExecTxnRunner`).
___
*[`pkg/sql/importer/import_job.go` line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDo8VmAfR0VjJgHxFp:-OYDo8VmAfR0VjJgHxFq:bdwziah) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L156)):*
> ```Go
> 	// expectedRowCountReadyByIndex maps from a BulkOpSummaryID defined by
> 	// the table id and the index id.
> 	// TODO(janexing): Using a map here could cause a data race but is it
> ```

Yeah, we should fix this. The data race is possible because we can start new goroutines via `grp.GoCtx` while still updating the map; instead, we should refactor the code so that we create the map in the first loop, and then in the second loop we'll start new goroutines.

Furthermore, we could simplify the code a bit `details.Tables` really can have only a single table now, so we could assert that we have exactly one table there, extract it and build `tblDesc` that will be utilized in both loops.
___
*[`pkg/sql/importer/import_job.go` line 206 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpGFC8X5qPW4V1kpg:-OYDpGFC8X5qPW4V1kph:bsgy1ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L206)):*
> ```Go
> 		// TODO(janexing): more graceful shutdown of pre-counts.
> 		for pendingPreCountTasks.Load() > 0 {
> 			time.Sleep(500 * time.Millisecond)
> ```

Hm, sleeping is not great, we should use some cleaner synchronization primitive to achieve what we need.

One idea would be to have a separate wait group that the main goroutine would `Add(numIndexes)` and then each worker goroutine would call `Done` after the pre-count is done. We'll need to be careful to always decrement the wait group in the worker (in erroneous cases like context cancellation or a crash).
___
*[`pkg/sql/importer/import_job.go` line 351 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpgmM7rlouW0-qS1O:-OYDpgmM7rlouW0-qS1P:bmdyqmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L351)):*
> ```Go
> 		// Signal row count validation goroutines with the expected count
> 		// from import.
> 		if expected, ok := expectedRowCountReadyByIndex[id]; ok {
> ```

Can `ok` be ever `false` here? What would that mean?
___
*[`pkg/sql/importer/import_job.go` line 357 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpsITCbGtXhdGsFVK:-OYDpsITCbGtXhdGsFVL:bhuju9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L357)):*
> ```Go
> 	}
> 
> 	rowCountValidationTestingKnob := r.testingKnobs.rowCountValidation
> ```

nit: pull the knob inside the `if`.
___
*[`pkg/sql/importer/import_job.go` line 359 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpyaHB-8AY-OmBb40:-OYDpyaHB-8AY-OmBb41:b-qnhjlf) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L359)):*
> ```Go
> 	rowCountValidationTestingKnob := r.testingKnobs.rowCountValidation
> 	if rowCountValidation {
> 		// TODO(janexing): shall we wait with timeout?
> ```

I don't think that's needed but it'd be worth double-checking that context cancellation is respected by the workers, so that if the IMPORT job as a whole is canceled, the row count validation reacts promptly.
___
*[`pkg/sql/importer/import_job.go` line 368 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDqBsHA3yeZSXRSzEJ:-OYDqBsHA3yeZSXRSzEK:bxjuwzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L368)):*
> ```Go
> 		log.Event(ctx, "row count validation completed successfully")
> 		if rowCountValidationTestingKnob != nil {
> 			close(rowCountValidationTestingKnob)
> ```

nit: should it be closed in case of an error?
___
*[`pkg/sql/importer/import_stmt_test.go` line 5759 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDqTS31zJW6VSrbI4A:-OYDqTS31zJW6VSrbI4B:b-nybyg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_stmt_test.go#L5759)):*
> ```Go
> 					return err
> 				case <-time.After(5 * time.Second):
> 					return errors.AssertionFailedf("timed out waiting for row count validation result")
> ```

nit: can this timeout be hit under heavy test configs like deadlock, in normal operation?


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
Sorry for the delayed response -- distracted by sub-issues (https://github.com/cockroachdb/cockroach/issues/152543 and https://github.com/cockroachdb/cockroach/issues/152519) revealed by this change. 

The failing CI test, `TestImportDefaultWithResume`, is pending fixes for these 2 issues, and the fix might be non-trivial. I'd vote for turning the validation off for this particular test, and open another PR dedicated for it. WDYT? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-:-OYl9lJS4LXk4Pwb2KTP:b-x0r4d5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @yuzefovich)

___
*[`pkg/sql/importer/import_job.go` line 126 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDm72I9fXRwJCJum3B:-OYHn8OfA6dMXsoUS8Ty:bl2ummh) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L126)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why not send the expected row count on the channel? Then we can avoid creating this utility struct.
</blockquote></details>

Good idea, changed it to a `chan int64`.
___
*[`pkg/sql/importer/import_job.go` line 134 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDnmST34Bed4pXu304:-OYHns2OAbmVJo-N1cNS:b6zrrk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L134)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We should probably use `now` instead of `p.ExecCfg().Clock.Now()` so that `now` is used in two places (similar to other callsites of `descs.NewHistoricalInternalExecTxnRunner`).
</blockquote></details>

Good catch!
___
*[`pkg/sql/importer/import_job.go` line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDo8VmAfR0VjJgHxFp:-OYl7ZhQ3X2mqq7Lggox:b7gnp2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L156)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yeah, we should fix this. The data race is possible because we can start new goroutines via `grp.GoCtx` while still updating the map; instead, we should refactor the code so that we create the map in the first loop, and then in the second loop we'll start new goroutines.

Furthermore, we could simplify the code a bit `details.Tables` really can have only a single table now, so we could assert that we have exactly one table there, extract it and build `tblDesc` that will be utilized in both loops.
</blockquote></details>

Hmm, the data race seems to be between the "write" at the count-before-import stage, and the "read" when import job finishes. I updated it to a sync.Map to save some headache. 

I also made a separate commit for asserting len(tables) == 1. Can merge it with the main commit too, no strong preference here. 
___
*[`pkg/sql/importer/import_job.go` line 206 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpGFC8X5qPW4V1kpg:-OYl8Da_4jpE20Nd5smL:b-qwa9cc) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L206)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, sleeping is not great, we should use some cleaner synchronization primitive to achieve what we need.

One idea would be to have a separate wait group that the main goroutine would `Add(numIndexes)` and then each worker goroutine would call `Done` after the pre-count is done. We'll need to be careful to always decrement the wait group in the worker (in erroneous cases like context cancellation or a crash).
</blockquote></details>

Good idea, change it to use wait group now.
___
*[`pkg/sql/importer/import_job.go` line 351 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpgmM7rlouW0-qS1O:-OYHtvKz7Z8pal7kAy3s:b-q31wsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L351)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Can `ok` be ever `false` here? What would that mean?
</blockquote></details>

Good catch! Yes, `ok` can be false, though it's a relatively unlikely scenario in the current flow.

The current execution flow is:

1.  **Precheck stage**: Count rows for each index and store index tags in `expectedRowCountReadyByIndex`
2.  **Set table offline**
3.  **Import data**
4.  **Set table public**
5.  **Post-import validation**: Signal validation goroutines with actual counts from `res.EntryCounts`
If `ok == false`, it means an index exists in `res.EntryCounts` (i.e., data was imported into it) but wasn't tracked during the precheck stage. This could happen if a new index is added after precheck is initiated but before import happens. Given this window is very short (between steps 1-3), it's unlikely but the check doesn't hurt.

However, I'm more concerned about the opposite scenario: what if an existing index is **dropped** during this window? In that case, we'd have a validation goroutine permanently waiting for a signal that will never come, since the dropped index won't appear in `res.EntryCounts`.

The fundamental issue is that we need precheck to happen before "bring table offline" takes effect, creating this race condition window. One potential solution would be to run pre-import counting via direct `p.ExecCfg().DB.Scan()` operations rather than calling `sql.CountIndexRowsAndMaybeCheckUniqueness()` (which query the count via internal executor) , which would eliminate the dependency on the table being online during counting. However, I need to look more into if counting via `p.ExecCfg().DB.Scan()` would be even slower, or is it safe for concurrent usage. WDYT?
___
*[`pkg/sql/importer/import_job.go` line 359 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpyaHB-8AY-OmBb40:-OYl8K76-BOGpP8gH5TC:bg2j5h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L359)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I don't think that's needed but it'd be worth double-checking that context cancellation is respected by the workers, so that if the IMPORT job as a whole is canceled, the row count validation reacts promptly.
</blockquote></details>

Makes sense, thanks!
___
*[`pkg/sql/importer/import_job.go` line 368 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDqBsHA3yeZSXRSzEJ:-OYHv0lf1PfN_yc_Uv9j:b5tm7b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L368)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should it be closed in case of an error?
</blockquote></details>

Good catch, added the close too. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice finds! Yeah, disabling the new validation in a single test sounds good to me - no need for this patch to wait for fixes to land.

> @yuzefovich reviewed 10 of 10 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-:-OYnf293EmxvfZ7lujJY:bwmhogu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`pkg/sql/importer/import_job.go` line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDo8VmAfR0VjJgHxFp:-OYnjhHh6kv1_YLpHqZv:bn06rsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L156)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Hmm, the data race seems to be between the "write" at the count-before-import stage, and the "read" when import job finishes. I updated it to a sync.Map to save some headache. 

I also made a separate commit for asserting len(tables) == 1. Can merge it with the main commit too, no strong preference here.
</blockquote></details>

I still think we can simplify this further. Currently the new goroutines have two jobs - do the pre-import counting that must happen before the table is taken offline and then do the post-import counting that must happen after the table is taken back online. If we split these two concerns into separate goroutines, I think we can clarify the synchronization with the main goroutine.

How about the following sequence:
- in the first loop over all indexes in the main goroutine we
  - create the channel for each `expectedRowCountReadyByIndex[tableTag]` - this channel will be blocked on by the post-import goroutine
  - do `preCountWaitGroup.Add(1)` and then start the pre-import count goroutine for each index
     - that goroutine populates `preImportRowCount[tableTag]` (unless we already have it from the previous job attempt) and then calls `preCountWaitGroup.Done()`
- after the first loop we do `preCountWaitGroup.Wait()` - this ensures that all pre-import count goroutines are done before we proceed any further. (We probably don't need to do any special context handling here since the pre-import count goroutine should respect the cancellation already, and in the worst case we'll just wait until all counts are finished.)
- now in the second loop over all indexes we create another goroutine for each index
   - this post-import count goroutine is first blocked on its channel from `expectedRowCountReadyByIndex` (the map is guaranteed to be fully populated at this point)
   - once it gets the expected count, it does the post-import count and assertion.
- (after all other steps) the main goroutine sends the expected count to each post-import count goroutine and then blocks on the wait group for the post-import count goroutines.

Does this make sense? I feel like this approach would be cleaner and would reduce the number of synchronization primitives that are used.
___
*[`pkg/sql/importer/import_job.go` line 351 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDpgmM7rlouW0-qS1O:-OYnoYl3886eht6xgRWH:b-1hinke) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L351)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Good catch! Yes, `ok` can be false, though it's a relatively unlikely scenario in the current flow.

The current execution flow is:

1.  **Precheck stage**: Count rows for each index and store index tags in `expectedRowCountReadyByIndex`
2.  **Set table offline**
3.  **Import data**
4.  **Set table public**
5.  **Post-import validation**: Signal validation goroutines with actual counts from `res.EntryCounts`
If `ok == false`, it means an index exists in `res.EntryCounts` (i.e., data was imported into it) but wasn't tracked during the precheck stage. This could happen if a new index is added after precheck is initiated but before import happens. Given this window is very short (between steps 1-3), it's unlikely but the check doesn't hurt.

However, I'm more concerned about the opposite scenario: what if an existing index is **dropped** during this window? In that case, we'd have a validation goroutine permanently waiting for a signal that will never come, since the dropped index won't appear in `res.EntryCounts`.

The fundamental issue is that we need precheck to happen before "bring table offline" takes effect, creating this race condition window. One potential solution would be to run pre-import counting via direct `p.ExecCfg().DB.Scan()` operations rather than calling `sql.CountIndexRowsAndMaybeCheckUniqueness()` (which query the count via internal executor) , which would eliminate the dependency on the table being online during counting. However, I need to look more into if counting via `p.ExecCfg().DB.Scan()` would be even slower, or is it safe for concurrent usage. WDYT?
</blockquote></details>

Interesting. Indeed, it feels like this new pre-import stage where we need the table to be online to use SQL to count the number of entries in each index can introduce some subtle races, like the one you're describing. Using the KV operations like `DB.Scan()` can also be non-trivial in case of multiple column families (we would need to decode each KV to see whether it belongs to the new or old SQL row).

I wonder whether we should build the mechanism for allowing _internal_ queries against the table descriptors that is offline, would that be too hacky? Might be worth chatting with Foundations about this.
___
*[`pkg/sql/importer/import_job.go` line 122 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYndK-61dyi6v7sSX1H:-OYndK-61dyi6v7sSX1I:bcij5je) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L122)):*
> ```Go
> 		"undetected data corruption if the import process fails to import the expected number of rows.",
> 	true,
> 	settings.WithUnsafe,
> ```

nit: we should either mark the setting with `settings.WithPublic`, or keep it undocumented and omit from the release note. I don't have strong opinions either way, so up to you.
___
*[`pkg/sql/importer/import_job.go` line 143 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYndjLj0ZtvwvFCvbzw:-OYndjLj0ZtvwvFCvbzx:b-2t1mc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L143)):*
> ```Go
> 	details := r.job.Details().(jobspb.ImportDetails)
> 	if details.PreImportRowCount == nil {
> 		// TODO(janexing): is here even the right place to initialize this field?
> ```

The right place probably would be around `import_planning.go:766`, and you already did the initialization there. Is it insufficient? Maybe it's because the map contents are empty there so the serialized proto results in `nil`? Given that the IMPORT job can be retried maybe it's actually desirable that we would create a fresh map for each attempt?

Separately, why are we putting this map into the proto at all? This looks like a state that we don't need to preserve across retries.
___
*[`pkg/sql/importer/import_job.go` line 179 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYnm04z3F1RBmKDO-U_:-OYnm04z3F1RBmKDO-Ua:b-y7sh5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L179)):*
> ```Go
> 					var preImportIdxLen int64
> 					var err error
> 					// Skip pre-import row counting when resuming a paused job, as the counts
> ```

Oh, ok, so this partially answers my other question for why we need to preserve this.
___
*[`pkg/sql/importer/import_job.go` line 189 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYnmmok0XfqjbFLrmDJ:-OYnmmok0XfqjbFLrmDK:b-qgu4ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L189)):*
> ```Go
> 						preImportIdxLen = cnt
> 					} else {
> 						preImportIdxLen, err = sql.CountIndexRowsAndMaybeCheckUniqueness(ctx, tblDesc, idx, false /* withFirstMutationPublic */, newHistoricalInternalExecByTime(p.ExecCfg().Clock.Now()), sessiondata.NoSessionDataOverride)
> ```

Should we actually call `Clock.Now()` only once for one job attempt? Is it possible for the clock to move forward while we're processing the IMPORT?

Also, we have `ImportDetails.Walltime` - maybe we should be using that timestamp for all clock-related operations?

Relatedly, is it possible that somehow the concurrent INSERT into the table, before it's taken offline, is not included into the pre-import count, but it'll be included in the post-import count?
___
*[`pkg/jobs/jobspb/jobs.proto` line 717 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYnd-Bo7JvoqR-wBOP1:-OYnd-Bo7JvoqR-wBOP2:b-xv4hq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/jobs/jobspb/jobs.proto#L717)):*
> ```proto
>   ];
> 
>   map<uint64, int64> pre_import_row_count = 28;
> ```

nit: if we do keep this field, it deserves a comment describing what we map from and to.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152169)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/jobs/jobspb/jobs.proto` line 717 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYnd-Bo7JvoqR-wBOP1:-OYs3nsCBKAkC_EQzw9n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/jobs/jobspb/jobs.proto#L717)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: if we do keep this field, it deserves a comment describing what we map from and to.
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_job.go` line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYDo8VmAfR0VjJgHxFp:-OYs5Q_w0jw7X-OM15Be:bbyqghf) ([raw file](https://github.com/cockroachdb/cockroach/blob/075253bed53851b11b18a25c4a3938b942ad447c/pkg/sql/importer/import_job.go#L156)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I still think we can simplify this further. Currently the new goroutines have two jobs - do the pre-import counting that must happen before the table is taken offline and then do the post-import counting that must happen after the table is taken back online. If we split these two concerns into separate goroutines, I think we can clarify the synchronization with the main goroutine.

How about the following sequence:
- in the first loop over all indexes in the main goroutine we
  - create the channel for each `expectedRowCountReadyByIndex[tableTag]` - this channel will be blocked on by the post-import goroutine
  - do `preCountWaitGroup.Add(1)` and then start the pre-import count goroutine for each index
     - that goroutine populates `preImportRowCount[tableTag]` (unless we already have it from the previous job attempt) and then calls `preCountWaitGroup.Done()`
- after the first loop we do `preCountWaitGroup.Wait()` - this ensures that all pre-import count goroutines are done before we proceed any further. (We probably don't need to do any special context handling here since the pre-import count goroutine should respect the cancellation already, and in the worst case we'll just wait until all counts are finished.)
- now in the second loop over all indexes we create another goroutine for each index
   - this post-import count goroutine is first blocked on its channel from `expectedRowCountReadyByIndex` (the map is guaranteed to be fully populated at this point)
   - once it gets the expected count, it does the post-import count and assertion.
- (after all other steps) the main goroutine sends the expected count to each post-import count goroutine and then blocks on the wait group for the post-import count goroutines.

Does this make sense? I feel like this approach would be cleaner and would reduce the number of synchronization primitives that are used.
</blockquote></details>

Great idea, and thanks for the details explanation! I implemented your idea, only without a wait group but with 2 `ctxgroup.Group` -- one for pre-import and one for post-import. And simply wait for these 2 groups. I think this serves the same purpose as the WaitGroup idea you proposed, unless I'm missing something here. 
___
*[`pkg/sql/importer/import_job.go` line 122 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYndK-61dyi6v7sSX1H:-OYs3kPGAltlkDg5Vier:b-416p6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L122)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we should either mark the setting with `settings.WithPublic`, or keep it undocumented and omit from the release note. I don't have strong opinions either way, so up to you.
</blockquote></details>

Thanks! Added ``settings.WithPublic`.
___
*[`pkg/sql/importer/import_job.go` line 143 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYndjLj0ZtvwvFCvbzw:-OYs3pMl6magj2NOxUdZ:b-vip0ae) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L143)):*
> The right place probably would be around `import_planning.go:766`, and you already did the initialization there. Is it insufficient?

No it's not sufficient to have it in `import_planning.go:766` -- it's because `UnmarshalPayload()` gives you the payload with a nil `PreImportRowCount` in `Payload.Details`. 

But I figure to initialize it for IMPORT within `UnmarshalPayload()` might add unnecessary overhead to not-import jobs; besides, there're other examples whose field is initialized during the main logic execution, such as [here](https://github.com/cockroachdb/cockroach/blob/e3770b8d2034d7eeaa5095a3b2214da3f35b895e/pkg/sql/row/expr_walker.go#L332-L335). Also, essentially, `PreImportRowCount` is only written and read within this `importResumer.Resume()` function. 

So I guess it should be fine to have this map be initialized here. WDYT?
___
*[`pkg/sql/importer/import_job.go` line 179 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYnm04z3F1RBmKDO-U_:-OYs7LkTEY-8o7AnNCpU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L179)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Oh, ok, so this partially answers my other question for why we need to preserve this.
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_job.go` line 189 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152169#-OYnmmok0XfqjbFLrmDJ:-OYs6L113mq5MH4nxnO0:b-heszoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7824dcdc690b76b6caf44f8d29081c6d11e11048/pkg/sql/importer/import_job.go#L189)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Should we actually call `Clock.Now()` only once for one job attempt? Is it possible for the clock to move forward while we're processing the IMPORT?

Also, we have `ImportDetails.Walltime` - maybe we should be using that timestamp for all clock-related operations?

Relatedly, is it possible that somehow the concurrent INSERT into the table, before it's taken offline, is not included into the pre-import count, but it'll be included in the post-import count?
</blockquote></details>

Hmm, it's a good question for concurrent INSERTs. I think a clean-cut solution for such edge cases is indeed, as you mentioned in another comment, allowing internal row count queries to be executed when the table is brought offline. 

What about I bring this to a discussion with sql-foundation and see if such a bypass-offline method is feasible? 


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
import: enable validation while table is brought offline

We want to enable row count validation before and after the actual import
behavior happens, while without being intefered by other possible concurrent
data modification on the same table. Thus, we should count while the table
is in OFFLINE mode.

Release note: None

ZhouXing19 commented:
Thanks for reviewing this PR! 
Closing it as we plan to switch to the async INSPECT approach. See more internal discussion [here](https://cockroachlabs.slack.com/archives/C02DSDS9TM1/p1757439225277989?thread_ts=1756991224.633879&cid=C02DSDS9TM1).
